### PR TITLE
Provider - terminate agreement on unreachable requestor

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #1086

### CODE REVIEW

File: agent/provider/src/payments/payments.rs [hunk 1]
```diff
@@ -494,82 +501,106 @@ impl Handler<UpdateCost> for Payments {
...
-        if let Some(activity) = agreement.activities.get(&msg.invoice_info.activity_id) {
-            if let ActivityPayment::Running { .. } = activity {
+        return match agreement.activities.get(&msg.invoice_info.activity_id) {
+            Some(ActivityPayment::Running { .. }) => {
                 let payment_model = agreement.payment_model.clone();
                 let context = self.context.clone();
+
+                let update_interval = agreement.update_interval;
+                let last_debit_note = agreement.last_send_debit_note.clone();
+                let deadline = agreement.payment_deadline.clone();
```
Comment (tworec) [inline agent/provider/src/payments/payments.rs:524]: This is not a `deadline` (absolute point in time), but a `timeout` (relative -- duration). Name misled me to understand code.

File: agent/provider/src/payments/payments.rs [hunk 2]
```diff
@@ -494,82 +501,106 @@ impl Handler<UpdateCost> for Payments {
...
+                                    msg.invoice_info.agreement_id
+                                ))
+                                .ok();
+                        }
+                    } else {
+                        myself.agreements
+                            .get_mut(&msg.invoice_info.agreement_id)
+                            .map(|agreement| agreement.last_send_debit_note = Utc::now());
                     }
+
```
Comment (tworec) [inline agent/provider/src/payments/payments.rs:561]: Is the below comment and action still valid? Shouldn't we rather withhold sending debit notes when `BreakAgreeement` was already sent?

File: agent/provider/src/payments/payments.rs [hunk 3]
```diff
@@ -389,49 +391,54 @@ impl Handler<ActivityCreated> for Payments {
...
-            // Add activity to list and send debit note after update_interval.
-            agreement.add_created_activity(&msg.invoice_info.activity_id);
-            ctx.notify_later(msg, agreement.update_interval);
+        // Sending UpdateCost with last_debit_note: None will start new
+        // DebitNotes chain for this activity.
+        let msg = UpdateCost {
+            invoice_info: DebitNoteInfo {
+                agreement_id: msg.agreement_id.clone(),
+                activity_id: msg.activity_id.clone(),
+                payment_deadline: None, // Will be added in UpdateCost handler.
```
Comment (tworec) [inline agent/provider/src/payments/payments.rs:414]: Why not fill it here?

**Review by tworec — CHANGES_REQUESTED** [review]

File: agent/provider/src/payments/payments.rs (hunk 2 above)
Comment (nieznanysprawiciel) [reply agent/provider/src/payments/payments.rs:561]: I could do it, but it will happen anyway

File: agent/provider/src/payments/payments.rs (hunk 3 above)
Comment (nieznanysprawiciel) [reply agent/provider/src/payments/payments.rs:414]: `UpdateCost` is called later. Check `ctx.notify_later` a few lines below.

File: agent/provider/src/payments/agreement.rs [hunk 4]
```diff
@@ -100,6 +105,10 @@ impl AgreementPayment {
     }
 
     pub fn add_created_activity(&mut self, activity_id: &str) {
+        // Track ability to send DebitNotes from Activity start.
+        // Note: we have single timestamp for all activities.
+        self.last_send_debit_note = Utc::now();
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply agent/provider/src/payments/agreement.rs:110]: This field is updated later after each sent DebitNote.

File: agent/provider/src/payments/agreement.rs [hunk 5]
```diff
@@ -52,7 +52,7 @@ pub struct AgreementPayment {
     pub activities: HashMap<String, ActivityPayment>,
 
     pub update_interval: std::time::Duration,
-    pub payment_deadline: Option<chrono::Duration>,
+    pub payment_timeout: Option<chrono::Duration>,
```
Comment (tworec) [inline agent/provider/src/payments/agreement.rs:55]: 🎉 thx a lot

File: agent/provider/src/payments/payments.rs [hunk 6]
```diff
@@ -817,15 +846,12 @@ impl Handler<DeadlineElapsed> for Payments {
...
-            })
-            .map_err(|e| {
-                log::error!(
-                    "Failed to send BreakAgreement when deadline elapsed for [{}]. {}",
-                    msg.agreement_id,
-                    e
-                )
+                reason: BreakReason::DebitNotesDeadline(timeout),
             })
+            .log_err(&format!(
```
Comment (tworec) [inline agent/provider/src/payments/payments.rs:851]:
```suggestion
            .log_err_msg(&format!(
```

**Review by tworec — APPROVED** [review]

**Pull request merged**